src/stirling/source_connectors/socket_tracer/protocols/cql/stitcher.cc

```diff
@@ -372,77 +372,120 @@ StatusOr<Record> ProcessSolitaryResp(Frame* resp_frame) {
         if (record_status.ok()) {
           entries.push_back(record_status.ConsumeValueOrDie());
         } else {
           VLOG(1) << record_status.ToString();
           ++error_count;
         }
+      }
+    }
 
-        // Found a match, so remove both request and response.
-        // We don't remove request frames on the fly, however,
-        // because it could otherwise cause unnecessary churn/copying in the deque.
-        // This is due to the fact that responses can come out-of-order.
-        // Just mark the request as consumed, and clean-up when they reach the head of the queue.
-        // Note that responses are always head-processed, so they don't require this optimization.
-        found_match = true;
-        req_frame.consumed = true;
-        break;
+    auto pos = requests->find(stream_id);
+    if (pos == requests->end()) {
+      VLOG(1) << absl::Substitute("Could not find any requests for stream = $0", stream_id);
+      // if we don't find a matching request, we can't do anything with this response
+      // so clean it up
+      resp_frames.clear();
+      if (!kEventHandled) {
+        ++error_count;
       }
+      continue;
+    }
+
+    // we found a potential set of requests for this stream ID
+    std::deque<cass::Frame>& req_frames = pos->second;
+    std::deque<uint64_t> req_timestamps = std::deque<uint64_t>();
+    // get just the timestamps for matching with responses
+    for (cass::Frame& req_frame : req_frames) {
+      req_timestamps.push_back(req_frame.timestamp_ns);
     }
 
-    if (!found_match) {
-      VLOG(1) << absl::Substitute("Did not find a request matching the response. Stream = $0",
-                                  resp_frame.hdr.stream);
-      ++error_count;
+    uint64_t latest_resp_ts = 0;
+    // go through the responses for this stream ID and check for requests
+    for (cass::Frame& resp_frame : resp_frames) {
+      latest_resp_ts = resp_frame.timestamp_ns;
+      // This returns the first request timestamp that is JUST BEFORE the response timestamp
+      // Upper bound returns the first request timestamps GREATER than the response timestamp; we
+      // want to get the one right before
+      auto stream_it =
+          std::upper_bound(req_timestamps.begin(), req_timestamps.end(), resp_frame.timestamp_ns) -
+          1;
+      int req_index = stream_it - req_timestamps.begin();
+      // Responses should always have a more recent timestamp than the first request. If this
+      // condition is triggered we should not attempt to match this frame. Since responses are
+      // cleared during StitchFrames this will get cleaned up during the current iteration.
+      if (stream_it + 1 == req_timestamps.begin()) {
+        CTX_DCHECK(false) << "Unable to find request that is earlier than response: "
+                          << resp_frame.ToString();
+        continue;
+      }
+      cass::Frame& req_frame = req_frames[req_index];
```

Comment:
Does this mean we look at only one `req_frame`? I thought if there were multiple request frames with time earlier than the response then we would look at all of those.

Reply:
The current logic assumes a one-to-one pairing, where each response matches the most recent preceding request. This is because, within a single stream, we expect to match a response with the most recent request.

So for each response frame (in the `resp_deque`), we try to find the most recent corresponding request frame (in `req_deque`) with a timestamp that's just before the response's timestamp.